Marlin/src/sd/SdBaseFile.cpp

```diff
@@ -1125,41 +1125,43 @@ int8_t SdBaseFile::readDir(dir_t* dir, char* longFilename) {
-            longFilename[currentPos++] = utf16_ch & 0x007F;
-          else if (0 == (utf16_ch & 0xF800)) {                                  // Encode as 2-byte utf-8 char
-            longFilename[currentPos++] = 0xC0 | ((utf16_ch >> 6) & 0x1F);
-            longFilename[currentPos++] = 0x80 | (utf16_ch & 0x3F);
-          }
-          else {
-            #if LONG_FILENAME_CHARSIZE > 2                                      // Encode as 3-byte utf-8 char
-              longFilename[currentPos++] = 0xE0 | ((utf16_ch >> 12) & 0x0F);
-              longFilename[currentPos++] = 0xC0 | ((utf16_ch >> 6) & 0x3F);
-              longFilename[currentPos++] = 0xC0 | (utf16_ch & 0x3F);
-            #else                                                               // Encode as '_'
+        if (longFilename) {
+          // Convert filename from utf-16 to utf-8 as Marlin expects
+          #if LONG_FILENAME_CHARSIZE > 2
+            // Add warning for developers for currently not supported 3-byte cases (Conversion series of 2-byte
+            // codepoints to 3-byte in-place will break the rest of filename)
+            #error "Currently filename re-encoding is done in-place. It may break the remaining chars to use 3-byte codepoints."
+          #endif
+          uint16_t currentPos = 0;
+
+          LOOP_L_N(i, (LONG_FILENAME_LENGTH / 2)) {
+            uint16_t idx = i * 2; // This is fixed as FAT LFN always contain UTF-16LE encoding
+
+            uint16_t utf16_ch = longFilename[idx] | (longFilename[idx + 1] << 8);
+            if (0xD800 == (utf16_ch & 0xF800))                                    // Surrogate pair - encode as '_'
               longFilename[currentPos++] = '_';
-            #endif
+            else if (0 == (utf16_ch & 0xFF80))                                    // Encode as 1-byte utf-8 char
+              longFilename[currentPos++] = utf16_ch & 0x007F;
+            else if (0 == (utf16_ch & 0xF800)) {                                  // Encode as 2-byte utf-8 char
+              longFilename[currentPos++] = 0xC0 | ((utf16_ch >> 6) & 0x1F);
+              longFilename[currentPos++] = 0x80 | (utf16_ch & 0x3F);
+            }
+            else {
+              #if LONG_FILENAME_CHARSIZE > 2                                      // Encode as 3-byte utf-8 char
+                longFilename[currentPos++] = 0xE0 | ((utf16_ch >> 12) & 0x0F);
+                longFilename[currentPos++] = 0xC0 | ((utf16_ch >> 6) & 0x3F);
+                longFilename[currentPos++] = 0xC0 | (utf16_ch & 0x3F);
+              #else                                                               // Encode as '_'
+                longFilename[currentPos++] = '_';
+              #endif
+            }
+
+            if (0 == utf16_ch) break; // End of filename
           }
-
-          if (0 == utf16_ch) break; // End of filename
+          n = currentPos;
         }
-        return currentPos;
-      #else
-        return n;
       #endif
+      return n;
```

Comment:
This is the only other bit that really changed.

Reply:
Function returns readed length. In case of longFilename doesn't inited, seems we can safely return constant zero.

Reply:
I attempted to preserve the existing behavior. I don't know what any callers do with the return data, so I avoided changing it.